src/control/server/harness.go

```diff
@@ -186,22 +186,10 @@ func (h *EngineHarness) CallDrpc(ctx context.Context, method drpc.Method, body p
 	// the first one that is available to service the request.
 	// If the request fails, that error will be returned.
 	for _, i := range h.Instances() {
-		if !i.IsReady() {
-			if i.IsStarted() {
-				if err == nil {
-					err = errInstanceNotReady
-				}
-			} else {
-				if err == nil {
-					err = FaultDataPlaneNotStarted
-				}
-			}
-			continue
-		}
 		resp, err = i.CallDrpc(ctx, method, body)
 
 		switch errors.Cause(err) {
-		case errDRPCNotReady, FaultDataPlaneNotStarted:
+		case errEngineNotReady, errDRPCNotReady:
```

Comment:
I think that this subtle change is what's causing the MS Failover test problem. The logic in this method was intended to invoke the dRPC failure handlers when the MS Leader has no ranks available to service a request. Looking at the control logs, we can see that the MS Leader's rank is killed, but it never resigns leadership, so the PoolCreate RPC is just retried over and over until timeout. What's supposed to happen is that the PoolCreate goes to the current MS Leader after its rank is killed, the RPC fails and the dRPC failure handlers are invoked (including a ResignLeadership() call), and the PoolCreate is then serviced by the new MS Leader.

I'm not sure what the best approach here is -- basically you're going to want to distinguish between the following scenarios:

- Engine wasn't started yet (retryable)
- Engine was started, but there was a race and the dRPC connection wasn't ready yet (retryable)
- Engine was started, but the dRPC call failed (not retryable, return the error to the caller)

I suspect that the changes in this PR have eliminated detection of the last scenario, with the result that the MS Leader never fails over.

Reply:
Thanks Mike, let me try to fix

Reply:
So should harness.CallDrpc() specifically avoid checking if the instances are ready before calling instance.callDrpc()?

Reply:
I've updated implementation